src/adapter/src/coord/timestamp_oracle/metrics.rs

```diff
@@ -0,0 +1,159 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Prometheus monitoring metrics.
+
+use std::time::Instant;
+
+use mz_ore::metric;
+use mz_ore::metrics::{Counter, IntCounter, MetricsRegistry};
+use mz_ore::stats::histogram_seconds_buckets;
+use mz_postgres_client::metrics::PostgresClientMetrics;
+use prometheus::{CounterVec, Histogram, HistogramVec, IntCounterVec};
+
+/// Prometheus monitoring metrics for timestamp oracles.
+///
+/// Intentionally not Clone because we expect this to be passed around in an
+/// Arc.
+pub struct Metrics {
+    _vecs: MetricsVecs,
+
+    /// Metrics for [`TimestampOracle`].
+    pub oracle: OracleMetrics,
+
+    /// Metrics for [`PostgresClient`].
+    pub postgres_client: PostgresClientMetrics,
+}
+
+impl std::fmt::Debug for Metrics {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Metrics").finish_non_exhaustive()
+    }
+}
+
+impl Metrics {
+    /// Returns a new [Metrics] instance connected to the given registry.
+    pub fn new(registry: &MetricsRegistry) -> Self {
+        let vecs = MetricsVecs::new(registry);
+
+        Metrics {
+            oracle: vecs.oracle_metrics(),
+            postgres_client: PostgresClientMetrics::new(registry, "mz_ts_oracle"),
+            _vecs: vecs,
+        }
+    }
+}
+
+#[derive(Debug)]
+struct MetricsVecs {
+    external_op_started: IntCounterVec,
+    external_op_succeeded: IntCounterVec,
+    _external_op_failed: IntCounterVec,
+    external_op_seconds: CounterVec,
+    external_op_latency: HistogramVec,
+}
+
+impl MetricsVecs {
+    fn new(registry: &MetricsRegistry) -> Self {
+        MetricsVecs {
+            external_op_started: registry.register(metric!(
+                name: "mz_ts_oracle_started_count",
+                help: "count of oracle operations started",
+                var_labels: ["op"],
+            )),
+            external_op_succeeded: registry.register(metric!(
+                name: "mz_ts_oracle_succeeded_count",
+                help: "count of oracle operations succeeded",
+                var_labels: ["op"],
+            )),
+            _external_op_failed: registry.register(metric!(
```

Comment:
We're expecting this to be used once we add retries?

Reply:
yep!